File: src/main/java/no/echokarriere/backend/user/UserRepository.java
package no.echokarriere.backend.user;

import no.echokarriere.backend.configuration.CrudRepository;
import no.echokarriere.db.tables.records.UserRecord;
import org.jooq.DSLContext;
import org.springframework.stereotype.Repository;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static no.echokarriere.db.Tables.USER;
import static org.jooq.impl.DSL.row;

@Repository
public class UserRepository implements CrudRepository<UserEntity, UUID> {
    private final DSLContext dsl;

    public UserRepository(DSLContext dsl) {
        this.dsl = dsl;
    }

    private static UserEntity map(UserRecord it) {
        return it.into(UserEntity.class);
    }

    @Override
    public Optional<UserEntity> create(UserEntity entity) {
        return dsl.insertInto(USER)
                .columns(USER.ID, USER.NAME, USER.EMAIL, USER.PASSWORD, USER.ACTIVE, USER.TYPE)
                .values(entity.getId(), entity.getName(), entity.getEmail(), entity.getPassword(), entity.isActive(), entity.getUserType())
                .returning()
                .fetchOptional()
                .map(UserRepository::map);
    }

    @Override
    public Optional<UserEntity> update(UserEntity entity) {
        return dsl.update(USER)
                .set(
                        row(USER.NAME, USER.EMAIL, USER.ACTIVE, USER.TYPE, USER.MODIFIED_AT),
                        row(entity.getName(), entity.getEmail(), entity.isActive(), entity.getUserType(), OffsetDateTime.now())
                )
                .where(USER.ID.eq(entity.getId()))
                .returning()
                .fetchOptional()
                .map(UserRepository::map);
    }

    @Override
    public boolean delete(UUID id) {
        return dsl.delete(USER)
                .where(USER.ID.eq(id))
                .execute() == 1;
    }

    @Override
    public Optional<UserEntity> select(UUID id) {
        return dsl.selectFrom(USER)
                .where(USER.ID.eq(id))
                .fetchOptional()
                .map(UserRepository::map);
    }

    @Override
    public List<UserEntity> selectAll() {
        return dsl.selectFrom(USER)
                .fetch()
                .map(UserRepository::map);
    }
}

File: src/main/kotlin/no/echokarriere/user/UserRepository.kt
package no.echokarriere.user

import no.echokarriere.Tables.USER
import no.echokarriere.configuration.Argon2Configuration
import no.echokarriere.configuration.CrudRepository
import no.echokarriere.dbQuery
import org.jooq.DSLContext
import java.util.UUID

class UserRepository(private val argon: Argon2Configuration, private val jooq: DSLContext) :
    CrudRepository<UserEntity, UUID> {
    override suspend fun selectAll(): List<UserEntity> = dbQuery {
        jooq.select()
            .from(USER)
            .fetch()
            .into(UserEntity::class.java)
    }

    override suspend fun select(id: UUID): UserEntity? = dbQuery {
        jooq.select()
            .from(USER)
            .where(USER.ID.eq(id))
            .fetchOne()
            ?.into(UserEntity::class.java)
    }

    suspend fun selectByEmail(email: String): UserEntity? = dbQuery {
        jooq.select()
            .from(USER)
            .where(USER.EMAIL.eq(email))
            .fetchOne()
            ?.into(UserEntity::class.java)
    }

    override suspend fun insert(entity: UserEntity): UserEntity? = dbQuery {
        jooq.insertInto(USER)
            .columns(USER.ID, USER.NAME, USER.EMAIL, USER.PASSWORD, USER.ACTIVE, USER.TYPE, USER.CREATED_AT)
            .values(
                entity.id,
                entity.name,
                entity.email,
                argon.hash(entity.password.toCharArray()),
                entity.active,
                entity.type,
                entity.createdAt
            )
            .returning()
            .fetchOne()
            ?.into(UserEntity::class.java)
    }

    override suspend fun update(entity: UserEntity): UserEntity? {
        TODO("Not yet implemented")
    }

    override suspend fun delete(id: UUID): Boolean {
        TODO("Not yet implemented")
    }
}
